File: 560 Subarray Sum Equals K/560 Subarray Sum Equals K.go
func subarraySum(nums []int, k int) int {
    pre := make(map[int]int, len(nums)+1)
    sum := 0
    pre[0] = 1
    res := 0
    for i := range nums {
        sum += nums[i]
        if v, ok := pre[sum-k]; ok {
            res += v
        }
        pre[sum]++
    }
    return res
}

File: 560 Subarray Sum Equals K/560 Subarray Sum Equals K.py
"""
Given an array of integers nums and an integer k, return the total number of continuous subarrays whose sum equals to k.

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2

Example 2:

Input: nums = [1,2,3], k = 3
Output: 2
"""
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        d = collections.defaultdict(int)
        d[0] = 1 #for first subarray k-0=k
        tmp_sum = 0;res = 0
        for i in range(len(nums)):
            tmp_sum += nums[i]
            if tmp_sum - k in d:
                res += d[tmp_sum - k]
            d[tmp_sum] += 1
        return res
